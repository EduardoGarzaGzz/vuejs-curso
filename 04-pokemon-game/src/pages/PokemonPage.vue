<template>
  <h1 v-if="!pokemon">Espere por favor....</h1>
  <template v-else>
    <h1>Quien es este pokemon?</h1>
    <PokemonPicture :pokemon-id="pokemon.id" :show-pokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemonsArr" @selection="checkAnswer" />
  </template>

  <template v-if="showAnswer" class="fade-in">
    <h2>{{ message }}</h2>
    <button @click="newGame">Nuevo juego</button>
  </template>
</template>

<script>
import PokemonPicture    from "@/components/PokemonPicture";
import PokemonOptions    from "@/components/PokemonOptions";
import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
  components: { PokemonOptions, PokemonPicture },
  data() {
    return {
      pokemonsArr: [],
      pokemon    : null,
      showPokemon: false,
      showAnswer : false,
      message    : ''
    }
  },
  methods: {
    async mixPokemonArray() {
      const rndInt = Math.floor( Math.random() * 4 );

      this.pokemonsArr = await getPokemonOptions();
      this.pokemon     = this.pokemonsArr[ rndInt ];
    },
    checkAnswer( pokemonId ) {
      this.showPokemon = true;
      this.showAnswer  = true;

      if ( pokemonId === this.pokemon.id ) {
        this.message = `Correcto, ${ this.pokemon.name }`;
      } else {
        this.message = `Oops, era ${ this.pokemon.name }`;
      }
    },
    newGame() {
      this.pokemon     = null;
      this.showPokemon = false;
      this.showAnswer  = false;
      this.message     = '';

      this.mixPokemonArray();
    }
  },
  mounted() {
    this.mixPokemonArray();
  }
}
</script>

<style>

</style>
