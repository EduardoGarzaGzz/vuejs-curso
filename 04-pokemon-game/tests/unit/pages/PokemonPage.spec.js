import { mount, shallowMount } from "@vue/test-utils";
import PokemonPage             from "@/pages/PokemonPage";
import { pokemons }            from "../mock/pokemons.mock";
import PokemonPicture          from "@/components/PokemonPicture";
import PokemonOptions          from "@/components/PokemonOptions";

describe( 'PokemonPage.spec.js page component', () => {
	let wrapper = null;
	
	beforeEach( () => {
		wrapper = shallowMount( PokemonPage );
	} );
	
	it( 'debe de hacer match con el snapshot', () => {
		expect( wrapper.html() ).toMatchSnapshot();
	} );
	
	it( 'debe de llamar mixPokemonArray al montar', () => {
		const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' );
		
		wrapper = shallowMount( PokemonPage );
		
		expect( mixPokemonArraySpy ).toHaveBeenCalled();
	} );
	
	it( 'debe de hacer match con el snapshot cuando cargan los pokemons', () => {
		wrapper = mount( PokemonPage, {
			data() {
				return {
					pokemonsArr: pokemons,
					pokemon    : pokemons[ 0 ],
					showPokemon: false,
					showAnswer : false,
					message    : ''
				}
			}
		} );
		
		expect( wrapper.html() ).toMatchSnapshot();
	} );
	
	it( 'debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
		wrapper = mount( PokemonPage, {
			data() {
				return {
					pokemonsArr: pokemons,
					pokemon    : pokemons[ 0 ],
					showPokemon: false,
					showAnswer : false,
					message    : ''
				}
			}
		} );
		
		const pokemonPictureWrapper = wrapper.findComponent( PokemonPicture )
		const pokemonOptionsWrapper = wrapper.findComponent( PokemonOptions )
		
		expect( pokemonPictureWrapper.exists() ).toBeTruthy();
		expect( pokemonOptionsWrapper.exists() ).toBeTruthy();
		
		expect( pokemonPictureWrapper.vm.pokemonId ).toBe( 5 );
		expect( pokemonOptionsWrapper.vm.pokemons ).not.toBe( [] );
	} );
	
	it( 'pruebas con checkAnswer', async () => {
		wrapper = shallowMount( PokemonPage, {
			data() {
				return {
					pokemonsArr: pokemons,
					pokemon    : pokemons[ 0 ],
					showPokemon: false,
					showAnswer : false,
					message    : ''
				}
			}
		} );
		
		await wrapper.vm.checkAnswer( 5 );
		expect( wrapper.find( 'h2' ).exists() ).toBeTruthy();
		expect( wrapper.vm.showPokemon ).toBe( true );
		expect( wrapper.find( 'h2' ).text() ).toBe( `Correcto, ${ pokemons[ 0 ].name }` );
		
		await wrapper.vm.checkAnswer( 10 );
		expect( wrapper.vm.message ).toBe( `Oops, era ${ pokemons[ 0 ].name }` );
	} );
} );
