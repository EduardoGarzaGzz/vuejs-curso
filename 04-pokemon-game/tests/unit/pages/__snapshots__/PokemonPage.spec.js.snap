// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage.spec.js page component debe de hacer match con el snapshot 1`] = `
<h1>Espere por favor....</h1>
<!--v-if-->
`;

exports[`PokemonPage.spec.js page component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<h1>Quien es este pokemon?</h1>
<div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/5.svg" alt="Un pokemon :D" class="hidden-pokemon">
  <!--v-if-->
</div>
<div class="options-container">
  <ul>
    <li>pickachu</li>
    <li>charmander</li>
    <li>venusaur</li>
    <li>charmander</li>
  </ul>
</div>
<!--v-if-->
`;
